File: packages/dotnet/oauth/src/DeviceFlow.cs
using System.Net.Http.Headers;
using System.Text.Json;

namespace OAuthDo;

/// <summary>
/// Device flow authorization for OAuth.
/// Implements the device authorization grant flow.
/// </summary>
public class DeviceFlow : IDisposable
{
    private const string AuthUrl = "https://auth.apis.do/user_management/authorize_device";
    private const string TokenUrl = "https://auth.apis.do/user_management/authenticate";
    public const string DefaultClientId = "client_01JQYTRXK9ZPD8JPJTKDCRB656";

    private readonly string _clientId;
    private readonly HttpClient _httpClient;
    private bool _disposed;

    /// <summary>
    /// Creates a DeviceFlow with the default client ID.
    /// </summary>
    public DeviceFlow() : this(DefaultClientId) { }

    /// <summary>
    /// Creates a DeviceFlow with a custom client ID.
    /// </summary>
    /// <param name="clientId">OAuth client ID</param>
    public DeviceFlow(string clientId)
    {
        _clientId = clientId;
        _httpClient = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(30)
        };
    }

    /// <summary>
    /// Initiates device authorization.
    /// </summary>
    /// <returns>Device authorization response</returns>
    public async Task<DeviceAuthResponse> AuthorizeAsync()
    {
        var content = new FormUrlEncodedContent(new[]
        {
            new KeyValuePair<string, string>("client_id", _clientId)
        });

        var response = await _httpClient.PostAsync(AuthUrl, content);
        var body = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
        {
            throw new DeviceFlowException($"Authorization failed: {body}");
        }

        using var json = JsonDocument.Parse(body);
        var root = json.RootElement;

        return new DeviceAuthResponse
        {
            DeviceCode = root.GetProperty("device_code").GetString() ?? "",
            UserCode = root.GetProperty("user_code").GetString() ?? "",
            VerificationUri = root.GetProperty("verification_uri").GetString() ?? "",
            VerificationUriComplete = root.TryGetProperty("verification_uri_complete", out var complete)
                ? complete.GetString()
                : null,
            ExpiresIn = root.TryGetProperty("expires_in", out var expires)
                ? expires.GetInt32()
                : 900,
            Interval = root.TryGetProperty("interval", out var interval)
                ? interval.GetInt32()
                : 5
        };
    }

    /// <summary>
    /// Polls for token after user authorization.
    /// </summary>
    /// <param name="deviceCode">Device code from authorization response</param>
    /// <param name="interval">Polling interval in seconds</param>
    /// <param name="timeout">Maximum time to wait in seconds</param>
    /// <param name="onPoll">Callback for each poll attempt (optional)</param>
    /// <param name="cancellationToken">Cancellation token</param>
    /// <returns>Token response</returns>
    public async Task<TokenResponse> PollForTokenAsync(
        string deviceCode,
        int interval = 5,
        int timeout = 900,
        Action<int>? onPoll = null,
        CancellationToken cancellationToken = default)
    {
        var startTime = DateTime.UtcNow;
        var pollInterval = interval;
        var attempts = 0;

        while ((DateTime.UtcNow - startTime).TotalSeconds < timeout)
        {
            cancellationToken.ThrowIfCancellationRequested();

            attempts++;
            onPoll?.Invoke(attempts);

            var content = new FormUrlEncodedContent(new[]
            {
                new KeyValuePair<string, string>("client_id", _clientId),
                new KeyValuePair<string, string>("device_code", deviceCode),
                new KeyValuePair<string, string>("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
            });

            var response = await _httpClient.PostAsync(TokenUrl, content, cancellationToken);
            var body = await response.Content.ReadAsStringAsync(cancellationToken);

            using var json = JsonDocument.Parse(body);
            var root = json.RootElement;

            if (root.TryGetProperty("error", out var errorProp))
            {
                var error = errorProp.GetString();
                var errorDescription = root.TryGetProperty("error_description", out var descProp)
                    ? descProp.GetString()
                    : null;

                if (error == "slow_down")
                {
                    pollInterval += 5;
                }

                if (error != "authorization_pending" && error != "slow_down")
                {
                    return new TokenResponse
                    {
                        Error = error,
                        ErrorDescription = errorDescription
                    };
                }
            }
            else
            {
                return new TokenResponse
                {
                    AccessToken = root.GetProperty("access_token").GetString(),
                    RefreshToken = root.TryGetProperty("refresh_token", out var refreshProp)
                        ? refreshProp.GetString()
                        : null,
                    TokenType = root.TryGetProperty("token_type", out var typeProp)
                        ? typeProp.GetString() ?? "Bearer"
                        : "Bearer",
                    ExpiresIn = root.TryGetProperty("expires_in", out var expiresProp)
                        ? expiresProp.GetInt32()
                        : 3600
                };
            }

            await Task.Delay(TimeSpan.FromSeconds(pollInterval), cancellationToken);
        }

        throw new TimeoutException($"Authorization timed out after {timeout} seconds");
    }

    /// <summary>
    /// Gets the client ID.
    /// </summary>
    /// <returns>OAuth client ID</returns>
    public string GetClientId() => _clientId;

    public void Dispose()
    {
        if (!_disposed)
        {
            _httpClient.Dispose();
            _disposed = true;
        }
    }
}

/// <summary>
/// Device authorization response.
/// </summary>
public class DeviceAuthResponse
{
    public string DeviceCode { get; set; } = string.Empty;
    public string UserCode { get; set; } = string.Empty;
    public string VerificationUri { get; set; } = string.Empty;
    public string? VerificationUriComplete { get; set; }
    public int ExpiresIn { get; set; } = 900;
    public int Interval { get; set; } = 5;

    public override string ToString() =>
        $"Visit {VerificationUri} and enter code: {UserCode}";
}

/// <summary>
/// Token response from authentication.
/// </summary>
public class TokenResponse
{
    public string? AccessToken { get; set; }
    public string? RefreshToken { get; set; }
    public string TokenType { get; set; } = "Bearer";
    public int ExpiresIn { get; set; } = 3600;
    public string? Error { get; set; }
    public string? ErrorDescription { get; set; }

    public bool IsError => Error is not null;
    public bool IsPending => Error is "authorization_pending" or "slow_down";
}

/// <summary>
/// Device flow exception.
/// </summary>
public class DeviceFlowException : Exception
{
    public DeviceFlowException(string message) : base(message) { }
}

File: packages/fsharp/oauth/src/DeviceFlow.fs
namespace OAuthDo

open System
open System.Net.Http
open System.Text.Json
open System.Threading

/// Device authorization response.
type DeviceAuthResponse = {
    DeviceCode: string
    UserCode: string
    VerificationUri: string
    VerificationUriComplete: string option
    ExpiresIn: int
    Interval: int
}

module DeviceAuthResponse =
    /// Creates a string representation of the response.
    let toString (response: DeviceAuthResponse) =
        sprintf "Visit %s and enter code: %s" response.VerificationUri response.UserCode

/// Token response from authentication.
type TokenResponse = {
    AccessToken: string option
    RefreshToken: string option
    TokenType: string
    ExpiresIn: int
    Error: string option
    ErrorDescription: string option
}

module TokenResponse =
    /// Checks if the response is an error.
    let isError (response: TokenResponse) = response.Error.IsSome

    /// Checks if the response is a pending authorization.
    let isPending (response: TokenResponse) =
        match response.Error with
        | Some "authorization_pending" | Some "slow_down" -> true
        | _ -> false

/// Device flow exception.
exception DeviceFlowException of string

/// Device flow authorization for OAuth.
/// Implements the device authorization grant flow.
type DeviceFlow(?clientId: string) =
    let authUrl = "https://auth.apis.do/user_management/authorize_device"
    let tokenUrl = "https://auth.apis.do/user_management/authenticate"
    let defaultClientId = "client_01JQYTRXK9ZPD8JPJTKDCRB656"

    let clientId = defaultArg clientId defaultClientId
    let httpClient = new HttpClient(Timeout = TimeSpan.FromSeconds(30.0))

    /// Gets the client ID.
    member _.ClientId = clientId

    /// Initiates device authorization.
    member _.AuthorizeAsync() = async {
        let content = new FormUrlEncodedContent([
            KeyValuePair("client_id", clientId)
        ])

        let! response = httpClient.PostAsync(authUrl, content) |> Async.AwaitTask
        let! body = response.Content.ReadAsStringAsync() |> Async.AwaitTask

        if not response.IsSuccessStatusCode then
            raise (DeviceFlowException(sprintf "Authorization failed: %s" body))

        use json = JsonDocument.Parse(body)
        let root = json.RootElement

        let getOptionalString (name: string) =
            match root.TryGetProperty(name) with
            | true, prop -> prop.GetString() |> Option.ofObj
            | false, _ -> None

        let getOptionalInt (name: string) (defaultValue: int) =
            match root.TryGetProperty(name) with
            | true, prop -> prop.GetInt32()
            | false, _ -> defaultValue

        return {
            DeviceCode = root.GetProperty("device_code").GetString()
            UserCode = root.GetProperty("user_code").GetString()
            VerificationUri = root.GetProperty("verification_uri").GetString()
            VerificationUriComplete = getOptionalString "verification_uri_complete"
            ExpiresIn = getOptionalInt "expires_in" 900
            Interval = getOptionalInt "interval" 5
        }
    }

    /// Polls for token after user authorization.
    member _.PollForTokenAsync(deviceCode: string, ?interval: int, ?timeout: int, ?onPoll: int -> unit, ?cancellationToken: CancellationToken) = async {
        let interval = defaultArg interval 5
        let timeout = defaultArg timeout 900
        let cancellationToken = defaultArg cancellationToken CancellationToken.None

        let startTime = DateTime.UtcNow
        let mutable pollInterval = interval
        let mutable attempts = 0
        let mutable result: TokenResponse option = None

        while result.IsNone && (DateTime.UtcNow - startTime).TotalSeconds < float timeout do
            cancellationToken.ThrowIfCancellationRequested()

            attempts <- attempts + 1
            onPoll |> Option.iter (fun f -> f attempts)

            let content = new FormUrlEncodedContent([
                KeyValuePair("client_id", clientId)
                KeyValuePair("device_code", deviceCode)
                KeyValuePair("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
            ])

            let! response = httpClient.PostAsync(tokenUrl, content, cancellationToken) |> Async.AwaitTask
            let! body = response.Content.ReadAsStringAsync(cancellationToken) |> Async.AwaitTask

            use json = JsonDocument.Parse(body)
            let root = json.RootElement

            let getOptionalString (name: string) =
                match root.TryGetProperty(name) with
                | true, prop -> prop.GetString() |> Option.ofObj
                | false, _ -> None

            let getOptionalInt (name: string) (defaultValue: int) =
                match root.TryGetProperty(name) with
                | true, prop -> prop.GetInt32()
                | false, _ -> defaultValue

            match root.TryGetProperty("error") with
            | true, errorProp ->
                let error = errorProp.GetString()
                let errorDescription = getOptionalString "error_description"

                if error = "slow_down" then
                    pollInterval <- pollInterval + 5

                if error <> "authorization_pending" && error <> "slow_down" then
                    result <- Some {
                        AccessToken = None
                        RefreshToken = None
                        TokenType = "Bearer"
                        ExpiresIn = 0
                        Error = Some error
                        ErrorDescription = errorDescription
                    }
            | false, _ ->
                result <- Some {
                    AccessToken = getOptionalString "access_token"
                    RefreshToken = getOptionalString "refresh_token"
                    TokenType = getOptionalString "token_type" |> Option.defaultValue "Bearer"
                    ExpiresIn = getOptionalInt "expires_in" 3600
                    Error = None
                    ErrorDescription = None
                }

            if result.IsNone then
                do! Async.Sleep(pollInterval * 1000)

        match result with
        | Some r -> return r
        | None -> return raise (TimeoutException(sprintf "Authorization timed out after %d seconds" timeout))
    }

    interface IDisposable with
        member _.Dispose() = httpClient.Dispose()
